Checker Fir

I would like to ask a question why your warnings do not work, for example, as in compose?
```Kotlin
@Poko
data class MyClass(
   val my: String
)
```
your errors are displayed after the launch!

I am asking this question purely out of curiosity, what is the reason for this. Thank you for the reply.

This requires K2 mode in the IDE, which is currently in beta and not enabled by default in the IDE. It also requires `kotlin.k2.only.bundled.compiler.plugins.enabled` to be disabled in the IDE registry; it is currently enabled by default.

Once these are set correctly on your machine, the FIR-reported errors _should_ work. But they don't – there's a Poko bug in the error reporting:
```
java.lang.NoClassDefFoundError: org/jetbrains/kotlin/com/intellij/psi/PsiElement
	at dev.drewhamilton.poko.fir.PokoFirCheckersExtension$Errors.<clinit>(PokoFirCheckersExtension.kt:138)
	at dev.drewhamilton.poko.fir.PokoFirCheckersExtension$PokoFirRegularClassChecker.check(PokoFirCheckersExtension.kt:49)
	at dev.drewhamilton.poko.fir.PokoFirCheckersExtension$PokoFirRegularClassChecker.check(PokoFirCheckersExtension.kt:36)
	at org.jetbrains.kotlin.fir.analysis.collectors.components.DeclarationCheckersDiagnosticComponent.check(DeclarationCheckersDiagnosticComponent.kt:133)
	at org.jetbrains.kotlin.fir.analysis.collectors.components.DeclarationCheckersDiagnosticComponent.visitRegularClass(DeclarationCheckersDiagnosticComponent.kt:57)
	at org.jetbrains.kotlin.fir.analysis.collectors.components.DeclarationCheckersDiagnosticComponent.visitRegularClass(DeclarationCheckersDiagnosticComponent.kt:19)
	at org.jetbrains.kotlin.fir.declarations.FirRegularClass.accept(FirRegularClass.kt:48)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.diagnostics.LLFirDiagnosticVisitor.checkElement(LLFirDiagnosticVisitor.kt:44)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.access$checkElement(AbstractDiagnosticCollectorVisitor.kt:30)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitWithDeclarationAndReceiver(AbstractDiagnosticCollectorVisitor.kt:1376)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitClassAndChildren(AbstractDiagnosticCollectorVisitor.kt:87)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitRegularClass(AbstractDiagnosticCollectorVisitor.kt:92)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitRegularClass(AbstractDiagnosticCollectorVisitor.kt:30)
	at org.jetbrains.kotlin.fir.declarations.FirRegularClass.accept(FirRegularClass.kt:48)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollector.collectDiagnostics(AbstractDiagnosticCollector.kt:36)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.diagnostics.FileStructureElementDiagnosticsCollector.collectForStructureElement(FileStructureElementDiagnosticsCollector.kt:29)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.diagnostics.FileStructureElementDiagnosticRetriever.retrieve(FileStructureElementDiagnosticRetriever.kt:45)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.diagnostics.FileStructureElementDiagnostics.diagnosticByCommonCheckers_delegate$lambda$0(FileStructureElementDiagnostics.kt:18)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.diagnostics.FileStructureElementDiagnostics.getDiagnosticByCommonCheckers(FileStructureElementDiagnostics.kt:17)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.diagnostics.FileStructureElementDiagnostics.access$getDiagnosticByCommonCheckers(FileStructureElementDiagnostics.kt:16)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure.collectDiagnosticsFromStructureElements(FileStructure.kt:244)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.file.structure.FileStructure.getAllDiagnosticsForFile(FileStructure.kt:143)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.diagnostics.DiagnosticsCollector.collectDiagnosticsForFile(DiagnosticsCollector.kt:25)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.state.LLSourceDiagnosticProvider.collectDiagnostics(LLDiagnosticProvider.kt:43)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.api.LLFirResolveSession.collectDiagnosticsForFile$low_level_api_fir(LLFirResolveSession.kt:94)
	at org.jetbrains.kotlin.analysis.low.level.api.fir.api.LowLevelFirApiFacadeKt.collectDiagnosticsForFile(LowLevelFirApiFacade.kt:87)
	at org.jetbrains.kotlin.analysis.api.fir.components.KaFirDiagnosticProvider.collectDiagnostics(KaFirDiagnosticProvider.kt:28)
	at org.jetbrains.kotlin.analysis.api.impl.base.KaBaseSession.collectDiagnostics(KaBaseSession.kt)
	at org.jetbrains.kotlin.idea.highlighting.KotlinDiagnosticHighlightVisitor.analyzeFile(KotlinDiagnosticHighlightVisitor.kt:64)
	at org.jetbrains.kotlin.idea.highlighting.KotlinDiagnosticHighlightVisitor.analyze(KotlinDiagnosticHighlightVisitor.kt:47)
	at com.intellij.codeInsight.daemon.impl.HighlightVisitorRunner.lambda$runVisitors$4(HighlightVisitorRunner.java:120)
	at com.intellij.concurrency.ApplierCompleter.processArrayItem(ApplierCompleter.java:116)
	at com.intellij.concurrency.ApplierCompleter.processArray(ApplierCompleter.java:196)
	at com.intellij.concurrency.ApplierCompleter.execAll(ApplierCompleter.java:166)
	at com.intellij.concurrency.JobLauncherImpl.lambda$invokeConcurrentlyUnderProgressAsync$3(JobLauncherImpl.java:103)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.tryRunReadAction(AnyThreadWriteThreadingSupport.kt:279)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:965)
	at com.intellij.concurrency.ApplierCompleter.lambda$wrapInReadActionAndIndicator$2(ApplierCompleter.java:148)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:660)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:735)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:691)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:659)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:79)
	at com.intellij.concurrency.ApplierCompleter.wrapInReadActionAndIndicator(ApplierCompleter.java:159)
	at com.intellij.concurrency.ApplierCompleter.lambda$wrapAndRun$1(ApplierCompleter.java:140)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.executeByImpatientReader(AnyThreadWriteThreadingSupport.kt:486)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:178)
	at com.intellij.concurrency.ApplierCompleter.wrapAndRun(ApplierCompleter.java:140)
	at com.intellij.concurrency.JobLauncherImpl.lambda$invokeConcurrentlyUnderProgressAsync$4(JobLauncherImpl.java:103)
	at com.intellij.concurrency.JobLauncherImpl.safeIterate(JobLauncherImpl.java:168)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgressAsync(JobLauncherImpl.java:100)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgress(JobLauncherImpl.java:48)
	at com.intellij.concurrency.JobLauncher.invokeConcurrentlyUnderProgress(JobLauncher.java:51)
	at com.intellij.codeInsight.daemon.impl.HighlightVisitorRunner.runVisitors(HighlightVisitorRunner.java:112)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.lambda$collectHighlights$15(GeneralHighlightingPass.java:247)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgressAsync(JobLauncherImpl.java:97)
	at com.intellij.concurrency.JobLauncherImpl.processConcurrentlyAsync(JobLauncherImpl.java:465)
	at com.intellij.codeInsight.daemon.impl.AnnotatorRunner.runAnnotatorsAsync(AnnotatorRunner.java:67)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.collectHighlights(GeneralHighlightingPass.java:257)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.lambda$collectInformationWithProgress$12(GeneralHighlightingPass.java:190)
	at com.intellij.codeInsight.daemon.impl.HighlightInfoUpdaterImpl.lambda$runWithInvalidPsiRecycler$32(HighlightInfoUpdaterImpl.java:689)
	at com.intellij.codeInsight.daemon.impl.ManagedHighlighterRecycler.runWithRecycler(ManagedHighlighterRecycler.java:87)
	at com.intellij.codeInsight.daemon.impl.HighlightInfoUpdaterImpl.runWithInvalidPsiRecycler(HighlightInfoUpdaterImpl.java:671)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.lambda$collectInformationWithProgress$13(GeneralHighlightingPass.java:215)
	at com.intellij.codeInsight.daemon.impl.HighlightVisitorRunner.createHighlightVisitorsFor(HighlightVisitorRunner.java:84)
	at com.intellij.codeInsight.daemon.impl.GeneralHighlightingPass.collectInformationWithProgress(GeneralHighlightingPass.java:112)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doCollectInformation(ProgressableTextEditorHighlightingPass.java:86)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:67)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:431)
	at com.intellij.platform.diagnostic.telemetry.helpers.TraceKt.runWithSpanIgnoreThrows(trace.kt:118)
	at com.intellij.platform.diagnostic.telemetry.helpers.TraceUtil.runWithSpanThrows(TraceUtil.java:36)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$3(PassExecutorService.java:426)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.tryRunReadAction(AnyThreadWriteThreadingSupport.kt:291)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:965)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$4(PassExecutorService.java:417)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:660)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:735)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:691)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:659)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:79)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:416)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:392)
	at com.intellij.openapi.fileTypes.impl.FileTypeManagerImpl.cacheFileTypesInside(FileTypeManagerImpl.java:802)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$1(PassExecutorService.java:392)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.executeByImpatientReader(AnyThreadWriteThreadingSupport.kt:486)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:178)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:390)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:259)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Caused by: java.lang.ClassNotFoundException: org.jetbrains.kotlin.com.intellij.psi.PsiElement
	at com.intellij.util.lang.UrlClassLoader.findClass(UrlClassLoader.java:233)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:593)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 94 more
```

Poko needs to switch to `K2Element` instead of `PsiElement`.

Finally, I can't get this to work in Android Studio, only IntelliJ. [This comment](https://issuetracker.google.com/issues/341233001#comment12) suggests AS should work too, so I'm not sure what I'm doing wrong there. I'll follow up there after fixing the first issue.

--- a/poko-compiler-plugin/src/main/kotlin/dev/drewhamilton/poko/fir/PokoFirCheckersExtension.kt
+++ b/poko-compiler-plugin/src/main/kotlin/dev/drewhamilton/poko/fir/PokoFirCheckersExtension.kt
@@ -1,12 +1,13 @@
 package dev.drewhamilton.poko.fir
 
 import org.jetbrains.kotlin.KtFakeSourceElementKind
-import org.jetbrains.kotlin.com.intellij.psi.PsiElement
 import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.diagnostics.AbstractSourceElementPositioningStrategy
+import org.jetbrains.kotlin.diagnostics.DiagnosticFactory0DelegateProvider
 import org.jetbrains.kotlin.diagnostics.DiagnosticReporter
 import org.jetbrains.kotlin.diagnostics.KtDiagnosticFactoryToRendererMap
+import org.jetbrains.kotlin.diagnostics.Severity
 import org.jetbrains.kotlin.diagnostics.SourceElementPositioningStrategies
-import org.jetbrains.kotlin.diagnostics.error0
 import org.jetbrains.kotlin.diagnostics.rendering.BaseDiagnosticRendererFactory
 import org.jetbrains.kotlin.diagnostics.rendering.RootDiagnosticRendererFactory
 import org.jetbrains.kotlin.diagnostics.reportOn
@@ -78,27 +79,40 @@ internal class PokoFirCheckersExtension(
 
     private object Errors : BaseDiagnosticRendererFactory() {
 
-        val PokoOnNonClass by error0<PsiElement>(
+        /**
+         * The compiler and the IDE use a different version of this class, so use reflection to find the available
+         * version.
+         */
+        // Adapted from https://github.com/TadeasKriz/K2PluginBase/blob/main/kotlin-plugin/src/main/kotlin/com/tadeaskriz/example/ExamplePluginErrors.kt#L8
+        private val psiElementClass by lazy {
+            try {
+                Class.forName("org.jetbrains.kotlin.com.intellij.psi.PsiElement")
+            } catch (_: ClassNotFoundException) {
+                Class.forName("com.intellij.psi.PsiElement")
+            }.kotlin
+        }
+
+        val PokoOnNonClass by error0(
             positioningStrategy = SourceElementPositioningStrategies.NAME_IDENTIFIER,
         )
 
-        val PokoOnDataClass by error0<PsiElement>(
+        val PokoOnDataClass by error0(
             positioningStrategy = SourceElementPositioningStrategies.DATA_MODIFIER,
         )
 
-        val PokoOnValueClass by error0<PsiElement>(
+        val PokoOnValueClass by error0(
             positioningStrategy = SourceElementPositioningStrategies.INLINE_OR_VALUE_MODIFIER,
         )
 
-        val PokoOnInnerClass by error0<PsiElement>(
+        val PokoOnInnerClass by error0(
             positioningStrategy = SourceElementPositioningStrategies.INNER_MODIFIER,
         )
 
-        val PrimaryConstructorRequired by error0<PsiElement>(
+        val PrimaryConstructorRequired by error0(
             positioningStrategy = SourceElementPositioningStrategies.NAME_IDENTIFIER,
         )
 
-        val PrimaryConstructorPropertiesRequired by error0<PsiElement>(
+        val PrimaryConstructorPropertiesRequired by error0(
             positioningStrategy = SourceElementPositioningStrategies.NAME_IDENTIFIER,
         )
 
@@ -131,6 +145,19 @@ internal class PokoFirCheckersExtension(
 
         init {
             RootDiagnosticRendererFactory.registerFactory(this)
+        }
+
+        /**
+         * Copy of [org.jetbrains.kotlin.diagnostics.error0] with hack for correct `PsiElement` class.
+         */
+        private fun error0(
+            positioningStrategy: AbstractSourceElementPositioningStrategy = SourceElementPositioningStrategies.DEFAULT,
+        ): DiagnosticFactory0DelegateProvider {
+            return DiagnosticFactory0DelegateProvider(
+                severity = Severity.ERROR,
+                positioningStrategy = positioningStrategy,
+                psiType = psiElementClass,
+            )
         }
     }
 }

